File: internal/cli/builtin_functions.go
package cli

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"runtime"
	"time"

	"github.com/PineappleBond/xyncra-server/pkg/client"
	"github.com/PineappleBond/xyncra-server/pkg/protocol"
)

// builtinFunctionInfos returns metadata for the built-in diagnostic
// functions that every client device automatically exposes (D-098, D-099).
func builtinFunctionInfos() []protocol.FunctionInfo {
	return []protocol.FunctionInfo{
		{
			Name:        "ping",
			Description: "Echo back a message with timestamp",
			Parameters: map[string]any{
				"type": "object",
				"properties": map[string]any{
					"message": map[string]any{
						"type":        "string",
						"description": "Message to echo back",
					},
				},
			},
			Tags: []string{"diagnostic"},
			Returns: &protocol.ReturnInfo{
				Type:        "object",
				Description: "Echoed message and server timestamp",
			},
		},
		{
			Name:        "get_device_info",
			Description: "Get basic device information",
			Parameters:  map[string]any{"type": "object"},
			Tags:        []string{"diagnostic"},
			Returns: &protocol.ReturnInfo{
				Type:        "object",
				Description: "Device hostname, OS, architecture, and process ID",
			},
		},
		{
			Name:        "get_time",
			Description: "Get current device time",
			Parameters:  map[string]any{"type": "object"},
			Tags:        []string{"diagnostic"},
			Returns: &protocol.ReturnInfo{
				Type:        "object",
				Description: "Current UTC time, unix timestamp, and timezone",
			},
		},
	}
}

// registerBuiltinHandlers registers the built-in function handlers on the
// given XyncraClient so the server can invoke them via reverse RPC (D-092).
func registerBuiltinHandlers(xc *client.XyncraClient) {
	xc.RegisterRequestHandler("ping", pingHandler)
	xc.RegisterRequestHandler("get_device_info", getDeviceInfoHandler)
	xc.RegisterRequestHandler("get_time", getTimeHandler)
}

// pingHandler handles "ping" requests: echoes the input message together
// with an RFC 3339 nano timestamp.
func pingHandler(_ context.Context, req *protocol.PackageDataRequest) (json.RawMessage, error) {
	var params struct {
		Message string `json:"message"`
	}
	if len(req.Params) > 0 {
		if err := json.Unmarshal(req.Params, &params); err != nil {
			return nil, fmt.Errorf("ping: parse params: %w", err)
		}
	}
	resp := map[string]any{
		"echo":      params.Message,
		"timestamp": time.Now().Format(time.RFC3339Nano),
	}
	data, err := json.Marshal(resp)
	if err != nil {
		return nil, fmt.Errorf("ping: marshal response: %w", err)
	}
	return data, nil
}

// getDeviceInfoHandler handles "get_device_info" requests: returns the
// host name, operating system, CPU architecture, and process ID.
func getDeviceInfoHandler(_ context.Context, _ *protocol.PackageDataRequest) (json.RawMessage, error) {
	hostname, err := os.Hostname()
	if err != nil {
		hostname = "unknown"
	}
	resp := map[string]any{
		"hostname": hostname,
		"os":       runtime.GOOS,
		"arch":     runtime.GOARCH,
		"pid":      os.Getpid(),
	}
	data, err := json.Marshal(resp)
	if err != nil {
		return nil, fmt.Errorf("get_device_info: marshal response: %w", err)
	}
	return data, nil
}

// getTimeHandler handles "get_time" requests: returns the current UTC time,
// Unix timestamp, and timezone string.
func getTimeHandler(_ context.Context, _ *protocol.PackageDataRequest) (json.RawMessage, error) {
	now := time.Now()
	resp := map[string]any{
		"utc":      now.UTC().Format(time.RFC3339Nano),
		"unix":     now.Unix(),
		"timezone": now.Location().String(),
	}
	data, err := json.Marshal(resp)
	if err != nil {
		return nil, fmt.Errorf("get_time: marshal response: %w", err)
	}
	return data, nil
}

File: demo/web/packages/xyncra-client-cli/src/builtin-functions.ts
/**
 * Built-in diagnostic functions automatically registered by every client.
 *
 * Mirrors Go internal/cli/builtin_functions.go exactly.
 *
 * @module
 */

import type { FunctionInfo, PackageDataRequest } from '@xyncra/protocol';
import type { XyncraClient, RequestHandlerFunc } from '@xyncra/client-core';
import { hostname } from 'node:os';
import { arch, platform } from 'node:os';

/**
 * Return metadata for the 3 built-in diagnostic functions.
 * These are auto-registered on every client device (D-098, D-099).
 */
export function builtinFunctionInfos(): FunctionInfo[] {
  return [
    {
      name: 'ping',
      description: 'Echo back a message with timestamp',
      parameters: {
        type: 'object',
        properties: {
          message: {
            type: 'string',
            description: 'Message to echo back',
          },
        },
      },
      tags: ['diagnostic'],
      returns: {
        type: 'object',
        description: 'Echoed message and server timestamp',
      },
    },
    {
      name: 'get_device_info',
      description: 'Get basic device information',
      parameters: { type: 'object' },
      tags: ['diagnostic'],
      returns: {
        type: 'object',
        description: 'Device hostname, OS, architecture, and process ID',
      },
    },
    {
      name: 'get_time',
      description: 'Get current device time',
      parameters: { type: 'object' },
      tags: ['diagnostic'],
      returns: {
        type: 'object',
        description: 'Current UTC time, unix timestamp, and timezone',
      },
    },
  ];
}

/**
 * Register built-in function handlers on the XyncraClient.
 * The server can invoke these via reverse RPC (D-092).
 */
export function registerBuiltinHandlers(client: XyncraClient): void {
  client.registerRequestHandler('ping', pingHandler);
  client.registerRequestHandler('get_device_info', getDeviceInfoHandler);
  client.registerRequestHandler('get_time', getTimeHandler);
}

/** Echo back the input message with a timestamp. */
async function pingHandler(req: PackageDataRequest): Promise<unknown> {
  const params = (req.params as Record<string, unknown>) ?? {};
  const message = (params.message as string) ?? '';
  return {
    echo: message,
    timestamp: new Date().toISOString(),
  };
}

/** Return device hostname, OS, architecture, and PID. */
async function getDeviceInfoHandler(_req: PackageDataRequest): Promise<unknown> {
  let host: string;
  try {
    host = hostname();
  } catch {
    host = 'unknown';
  }
  return {
    hostname: host,
    os: platform(),
    arch: arch(),
    pid: process.pid,
  };
}

/** Return current UTC time, Unix timestamp, and timezone. */
async function getTimeHandler(_req: PackageDataRequest): Promise<unknown> {
  const now = new Date();
  return {
    utc: now.toISOString(),
    unix: Math.floor(now.getTime() / 1000),
    timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
  };
}
